parser_test.go

```diff
@@ -2999,33 +3012,129 @@ func (s *testParserSuite) TestTablePartition(c *C) {
+		{"create table t1 (a int) partition by list (a) (partition x history, partition y current)", false, ""},
+		{"create table t1 (a int) partition by system_time (partition x history, partition y current)", true, "CREATE TABLE `t1` (`a` INT) PARTITION BY SYSTEM_TIME (PARTITION `x` HISTORY,PARTITION `y` CURRENT)"},
+
+		// LIST, RANGE and SYSTEM_TIME partitions all required definitions
+		{"create table t1 (a int) partition by hash (a)", true, "CREATE TABLE `t1` (`a` INT) PARTITION BY HASH (`a`) PARTITIONS 1"},
+		{"create table t1 (a int) partition by key (a)", true, "CREATE TABLE `t1` (`a` INT) PARTITION BY KEY (`a`) PARTITIONS 1"},
+		{"create table t1 (a int) partition by range (a)", false, ""},
+		{"create table t1 (a int) partition by list (a)", false, ""},
+		{"create table t1 (a int) partition by system_time", false, ""},
+		// SYSTEM_TIME required 2 or more partitions
+		{"create table t1 (a int) partition by system_time (partition x history)", false, ""},
+		{"create table t1 (a int) partition by system_time (partition x current)", false, ""},
+
+		// number of columns and number of values in VALUES clauses must match
+		{"create table t1 (a int, b int) partition by range (a) (partition x values less than (10, 20))", false, ""},
+		{"create table t (id int) partition by range columns (id) (partition p0 values less than (1, 2))", false, ""},
+		{"create table t1 (a int, b int) partition by range columns (a, b) (partition x values less than (10, 20))", true, "CREATE TABLE `t1` (`a` INT,`b` INT) PARTITION BY RANGE COLUMNS (`a`,`b`) (PARTITION `x` VALUES LESS THAN (10, 20))"},
+		{"create table t1 (a int, b int) partition by range columns (a, b) (partition x values less than (10))", false, ""},
+		{"create table t1 (a int, b int) partition by range columns (a, b) (partition x values less than maxvalue)", false, ""},
+		{"create table t1 (a int, b int) partition by list (a) (partition x values in ((10, 20)))", false, ""},
+		{"create table t1 (a int, b int) partition by list columns (a, b) (partition x values in ((10, 20)))", true, "CREATE TABLE `t1` (`a` INT,`b` INT) PARTITION BY LIST COLUMNS (`a`,`b`) (PARTITION `x` VALUES IN ((10, 20)))"},
+		{"create table t1 (a int, b int) partition by list columns (a, b) (partition x values in (10, 20))", false, ""},
+		{"create table t1 (a int, b int) partition by list columns (a, b) (partition x values in (10, (20, 30)))", false, ""},
+		{"create table t1 (a int, b int) partition by list columns (a, b) (partition x values in ((10, 20), 30))", false, ""},
+		{"create table t1 (a int, b int) partition by list columns (a, b) (partition x values in ((10, 20), (30, 40, 50)))", false, ""},
+
+		// there must be at least one column/partition/value inside (...)
+		{"create table t1 (a int) partition by hash (a) ()", false, ""},
+		{"create table t1 (a int primary key) partition by key ()", true, "CREATE TABLE `t1` (`a` INT PRIMARY KEY) PARTITION BY KEY () PARTITIONS 1"},
```

Comment:
@tiancaiamao Yes it is valid, this means partition by the primary key or a non-null unique key.

https://dev.mysql.com/doc/refman/8.0/en/partitioning-key.html

> Where no column name is specified as the partitioning key, the table's primary key is used, if there is one. 
>
>  If there is no primary key but there is a unique key, then the unique key is used for the partitioning key: 
>
> However, if the unique key column were not defined as `NOT NULL`, then the previous statement would fail.